[11ty] New README and other polish

aomarks opened the pull request:
- Adds a new README for `@lit-labs/eleventy-plugin-lit`.

  <img src="https://user-images.githubusercontent.com/48894/152409553-c7a5a354-53d8-4509-b7a0-1aeca1723c41.png" width="300">


- Adds a new useful error when `--experimental-vm-modules` flag was omitted (previously it threw with something like `vm.SourceTextModule is not a constructor`).

  <img src="https://user-images.githubusercontent.com/48894/152406887-fc53aed6-6b5b-479a-a10e-58e1d98b777e.png" width="400">

- Add Node engine `>=12.16.0` requirement.

- Fix a typo in the `@lit-labs/ssr` README.

- Link to `@lit-labs/eleventy-plugin-lit` directory from the top-level monorepo README.

kevinpschaaf commented on packages/labs/eleventy-plugin-lit/demo/_includes/default.html:
```diff
@@ -1,12 +1,56 @@
+        if (!HTMLTemplateElement.prototype.hasOwnProperty('shadowRoot')) {
+          // Fetch the declarative shadow DOM ponyfill.
+          const {hydrateShadowRoots} = await import(
+            '@webcomponents/template-shadowroot/template-shadowroot.js'
```
Not sure there's a way to avoid this, but I assume if you slow your network down on e.g. Safari/Firefox, you'd get fouc because there's no way to conditionally load this _and_ make it run sync before paint?

FYI if you haven't seen it, there's a mostly untested [mutation-observer](https://github.com/webcomponents/template-shadowroot/blob/master/src/_implementation/mutation_observer.ts) implementation of the polyfill that could make this more robust by loading it up front and letting it upgrade as the parser yields (at the cost of loading it unconditionally, even on Chrome). We wanted to perf-test all the options and figure out the most robust/fastest way to polyfill it (which is why it's in such a manual state right now), so this is probably another thing to put on the roadmap.

aomarks replied:
Yeah good point, even on my machine there is a FOUC on Firefox with the demo. I also was thinking of a pattern like:

```js
<style>
  body[dsd-pending] { display: none; }
</style>
<body dsd-pending>
  <script>
  if (HTMLTemplateElement.prototype.hasOwnProperty('shadowRoot')) {
    body.removeAttribute('dsd-pending');
  }
  </script>

  {{content}}

  <script type="module">
    ...
    if (!HTMLTemplateElement.prototype.hasOwnProperty('shadowRoot')) {
      ...
      hydrateShadowRoots(document.body);
      body.removeAttribute('dsd-pending');
    }
    ...
  </script>
</body>
```

The downside of this is that it wouldn't allow for any incremental painting before all DOM content is parsed, and it's a little complicated.

The `MutationObserver` implementation seems worth pursuing though, yeah. If it's small, inlining it in the `<head>` could turn out to be pretty optimal.

aomarks replied:
Filed https://github.com/lit/lit/issues/2490 and added it to the Roadmap section.

justinfagnani replied:
I like that pattern, it's very much like the old Polymer feature or `:not(:defined) { display: none; }`

There's also another ponyfill approach where the JS isloaded up front and we emit calls to attach shadow roots in inline scripts right after each template.

aomarks commented on packages/labs/eleventy-plugin-lit/README.md:
```diff
@@ -1,68 +1,354 @@
+        await litHydrateSupportInstalled;
+
+        // Load component definitions. As each component definition loads, your
+        // pre-rendered components will come to life and become interactive.
```
Done.

aomarks commented:
Added a disclaimer about component compatibility, with a link to an issue, as suggested offline by @graynorton. Also added some additional links in a few spots.

justinfagnani commented on packages/labs/eleventy-plugin-lit/README.md:
```diff
@@ -1,68 +1,396 @@
+<h1>Greetings</h1>
+
+<demo-greeter name="World">
+  <template shadowroot="open">
```
Maybe add a small note right below the sample that `<template shadowroot="open">` is an HTML standard for declarative shadow roots and either browser will automatically attach the shadow root (Chrome, Edge) or a polyfill will. With links.

aomarks replied:
Done, but just a brief sentence that links to the Declarative shadow DOM section that we already have below.

justinfagnani commented on packages/labs/eleventy-plugin-lit/README.md:
```diff
@@ -1,68 +1,396 @@
+There are currently a number of restrictions that determine whether a component
+will be compatible with Lit pre-rendering, because not all of the component
+lifecycle methods are currently invoked, and the DOM APIs that can be used in
+certain lifecycle methods are restricted.
```
I'd put the note after so it fits in the flow, or repeat it a bit: "The Lit team is working on finalizing and documenting the SSR lifecycle and restrictions."

aomarks replied:
Done.

justinfagnani commented on packages/labs/eleventy-plugin-lit/README.md:
```diff
@@ -1,68 +1,396 @@
+> discussion. 🚧
+
+Data can be passed to your components by setting attributes (see the `name`
+attribute in the [example](#usage) above).
```
Mention children?

aomarks replied:
You mean slotted children, correct?

aomarks replied:
Done.

justinfagnani commented on packages/labs/eleventy-plugin-lit/README.md:
```diff
@@ -1,68 +1,396 @@
+      rel="modulepreload"
+      href="/node_modules/lit/experimental-hydrate-support.js"
+    />
+    <link rel="modulepreload" href="/_js/component1.js" />
```
We should add helpers for this. Could we add an issue and a note?

aomarks replied:
Done (https://github.com/lit/lit/issues/2487 and https://github.com/lit/lit/issues/2490). Added a note.

justinfagnani commented on packages/labs/eleventy-plugin-lit/src/test/eleventy-plugin-test.ts:
```diff
@@ -108,22 +117,28 @@ test.after.each(async ({rig}) => {
-  }
-  throw lastError ?? new Error('Timed out immediately');
+  return new Promise<T>((resolve, reject) => {
+    const check = async () => {
```
What happened here? Did the while loop not work?

justinfagnani commented on packages/labs/eleventy-plugin-lit/README.md:
```diff
@@ -1,68 +1,430 @@
+
+Data can be passed to your components by setting attributes (see the `name`
+attribute in the [example](#usage) above), or by passing children into an
+element with a [`<slot>`](https://lit.dev/docs/components/shadow-dom/#slots).
```
The element doesn't even need a `<slot>` to use children as data. It can iterate, query, etc. over children and interact with events.
